File: MultiHotelManagementAngular/src/app/services/gallery.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Gallery, GalleryRequest } from '../models/gallery.model';
import { environment } from '../../environments/environments';

@Injectable({ providedIn: 'root' })
export class GalleryService {
  private API_URL = environment.apiUrl + 'gallery';

  constructor(private http: HttpClient) {}

  getAll() {
    return this.http.get<Gallery[]>(this.API_URL);
  }

  getById(id: number) {
    return this.http.get<Gallery>(`${this.API_URL}/${id}`);
  }

  getByHotel(hotelId: number) {
    return this.http.get<Gallery[]>(`${this.API_URL}/hotel/${hotelId}`);
  }

  getByCategory(hotelId: number, category: string) {
    return this.http.get<Gallery[]>(`${this.API_URL}/hotel/${hotelId}/category/${category}`);
  }

  create(data: GalleryRequest, image: File) {
    const formData = new FormData();
    formData.append('data', new Blob([JSON.stringify(data)], { type: 'application/json' }));
    formData.append('image', image);
    return this.http.post<Gallery>(`${this.API_URL}/upload`, formData);
  }

  uploadMultiple(hotelId: number, images: File[], category: string = 'GENERAL') {
    const formData = new FormData();
    images.forEach((img) => formData.append('images', img));
    return this.http.post<Gallery[]>(
      `${this.API_URL}/upload/multiple/${hotelId}?category=${category}`,
      formData,
    );
  }

  update(id: number, data: GalleryRequest) {
    return this.http.put<Gallery>(`${this.API_URL}/${id}`, data);
  }

  delete(id: number) {
    return this.http.delete<void>(`${this.API_URL}/${id}`);
  }

  deleteByHotel(hotelId: number) {
    return this.http.delete<void>(`${this.API_URL}/hotel/${hotelId}`);
  }

  deleteByCategory(hotelId: number, category: string) {
    return this.http.delete<void>(`${this.API_URL}/hotel/${hotelId}/category/${category}`);
  }
}

File: MultiHotelManagement/src/main/java/com/MHM/MultiHotelManagement/service/GalleryService.java
package com.MHM.MultiHotelManagement.service;

import com.MHM.MultiHotelManagement.dto.request.GalleryRequestDTO;
import com.MHM.MultiHotelManagement.dto.response.GalleryResponseDTO;
import org.springframework.web.multipart.MultipartFile;

import java.util.List;

public interface GalleryService {

    // Single image upload
    GalleryResponseDTO upload(
            GalleryRequestDTO dto,
            MultipartFile image
    );

    // Multiple images upload একসাথে
    List<GalleryResponseDTO> uploadMultiple(
            Long hotelId,
            String category,
            List<MultipartFile> images
    );

    // সব gallery
    List<GalleryResponseDTO> getAll();

    // ID দিয়ে খোঁজো
    GalleryResponseDTO getById(Long id);

    // Hotel ID দিয়ে সব image খোঁজো
    List<GalleryResponseDTO> getByHotelId(Long hotelId);

    // Hotel ID + Category দিয়ে খোঁজো
    List<GalleryResponseDTO> getByHotelIdAndCategory(
            Long hotelId,
            String category
    );

    // Caption বা Category update করো
    GalleryResponseDTO update(
            Long id,
            GalleryRequestDTO dto
    );

    // একটা image delete করো
    void delete(Long id);

    // Hotel এর সব image delete করো
    void deleteAllByHotelId(Long hotelId);

    // Hotel এর নির্দিষ্ট Category র সব image delete করো
    void deleteByHotelIdAndCategory(
            Long hotelId,
            String category
    );
}